File: examples/showcase/gitlab_fronting/custom-code-bundles/gitlab_fronting_csharp/BackendAdapter.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using Anip.Service;

namespace {{ANIP_CSHARP_ROOT_NAMESPACE}};

public delegate Dictionary<string, object?> BackendAdapterHandler(
    Dictionary<string, object?> capability,
    Dictionary<string, object?> plan,
    Dictionary<string, object?> adapterInput,
    InvocationContext context);

public static class BackendAdapter
{
    public static BackendAdapterHandler Default => Execute;

    private static readonly HttpClient Http = new();

    private static Dictionary<string, object?> Execute(
        Dictionary<string, object?> capability,
        Dictionary<string, object?> plan,
        Dictionary<string, object?> parameters,
        InvocationContext context)
    {
        var unresolved = StringList(plan.GetValueOrDefault("unresolved_required_backend_inputs"));
        if (unresolved.Count > 0)
        {
            return Result(capability, plan, "backend_input_incomplete", new() { ["unresolved_required_backend_inputs"] = unresolved });
        }

        var token = Environment.GetEnvironmentVariable("GITLAB_TOKEN")?.Trim() ?? "";
        if (string.IsNullOrWhiteSpace(token))
        {
            return Result(capability, plan, "backend_error", new() { ["gitlab_error"] = new Dictionary<string, object?> { ["error"] = "missing_gitlab_token" } });
        }

        return Text(capability.GetValueOrDefault("capability_id")) switch
        {
            "gitlab.project.search_context" => SearchProjectContext(capability, plan, parameters, token),
            "gitlab.issue.prepare" => PrepareOrCreateIssue(capability, plan, parameters, token, context),
            "gitlab.mr.comment.prepare" => PrepareMergeRequestComment(capability, plan, parameters, token, context),
            "gitlab.pipeline.trigger.request" => PreparePipelineTrigger(capability, plan, parameters, token),
            "gitlab.release_notes.prepare" => PrepareReleaseNotes(capability, plan, parameters, token),
            _ => Result(capability, plan, "backend_execution_stub", new() { ["note"] = "No GitLab custom handler is registered for this capability." }),
        };
    }

    private static Dictionary<string, object?> SearchProjectContext(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, string token)
    {
        var project = ProjectId(parameters);
        if (!ProjectAllowed(project)) return Restricted(capability, plan, project);
        var limit = BoundedLimit(parameters.GetValueOrDefault("limit"), 20, 50);
        var query = Text(parameters.GetValueOrDefault("query"));
        var issues = GitLab("GET", $"/projects/{Uri.EscapeDataString(project)}/issues?search={Uri.EscapeDataString(query)}&per_page={limit}", token, null);
        var mergeRequests = GitLab("GET", $"/projects/{Uri.EscapeDataString(project)}/merge_requests?search={Uri.EscapeDataString(query)}&per_page={limit}", token, null);
        if (issues.ContainsKey("error")) return Result(capability, plan, "backend_error", new() { ["gitlab_error"] = issues });
        if (mergeRequests.ContainsKey("error")) return Result(capability, plan, "backend_error", new() { ["gitlab_error"] = mergeRequests });
        var items = new List<Dictionary<string, object?>>();
        foreach (var item in JsonList(issues.GetValueOrDefault("items")))
        {
            if (items.Count >= limit) break;
            items.Add(new() { ["kind"] = "issue", ["iid"] = Get(item, "iid"), ["title"] = Get(item, "title"), ["state"] = Get(item, "state"), ["web_url"] = Get(item, "web_url") });
        }
        foreach (var item in JsonList(mergeRequests.GetValueOrDefault("items")))
        {
            if (items.Count >= limit) break;
            items.Add(new() { ["kind"] = "merge_request", ["iid"] = Get(item, "iid"), ["title"] = Get(item, "title"), ["state"] = Get(item, "state"), ["web_url"] = Get(item, "web_url") });
        }
        return Result(capability, plan, "completed", new() { ["gitlab_query"] = query, ["result"] = new Dictionary<string, object?> { ["items"] = items, ["count"] = items.Count, ["project_id"] = project } });
    }

    private static Dictionary<string, object?> PrepareOrCreateIssue(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, string token, InvocationContext context)
    {
        var project = ProjectMetadata(capability, plan, parameters, token);
        if (project.Error is not null) return project.Error;
        var body = new Dictionary<string, object?> { ["title"] = Text(parameters.GetValueOrDefault("title")), ["description"] = FirstNonEmpty(Text(parameters.GetValueOrDefault("body")), Text(parameters.GetValueOrDefault("description"))) };
        var labels = StringList(parameters.GetValueOrDefault("labels"));
        if (labels.Count > 0) body["labels"] = string.Join(",", labels);
        var preview = WritePreview(capability, plan, "issues.create", $"/projects/{project.Id}/issues", body, project.Payload!);
        if (!MutationEnabled(context)) return preview;
        var created = GitLab("POST", $"/projects/{Uri.EscapeDataString(project.Id)}/issues", token, body);
        if (created.ContainsKey("error"))
        {
            preview["execution_status"] = "backend_error";
            preview["gitlab_error"] = created;
            return preview;
        }
        preview["execution_status"] = "completed";
        preview["approval_required"] = false;
        preview["mutation_performed"] = true;
        preview["created_issue"] = new Dictionary<string, object?> { ["iid"] = created.GetValueOrDefault("iid"), ["web_url"] = created.GetValueOrDefault("web_url"), ["state"] = created.GetValueOrDefault("state") };
        preview["note"] = "Created GitLab issue after the ANIP runtime validated and reserved an approval grant.";
        return preview;
    }

    private static Dictionary<string, object?> PrepareMergeRequestComment(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, string token, InvocationContext context)
    {
        var project = ProjectMetadata(capability, plan, parameters, token);
        if (project.Error is not null) return project.Error;
        var iid = Text(parameters.GetValueOrDefault("merge_request_iid"));
        var mergeRequest = GitLab("GET", $"/projects/{Uri.EscapeDataString(project.Id)}/merge_requests/{Uri.EscapeDataString(iid)}", token, null);
        if (mergeRequest.ContainsKey("error")) return Result(capability, plan, "backend_error", new() { ["gitlab_error"] = mergeRequest });
        var body = new Dictionary<string, object?> { ["body"] = $"[{FirstNonEmpty(Text(parameters.GetValueOrDefault("comment_purpose")), "triage_update")}] {Text(parameters.GetValueOrDefault("context"))}".Trim() };
        var preview = WritePreview(capability, plan, "merge_requests.createNote", $"/projects/{project.Id}/merge_requests/{iid}/notes", body, project.Payload!);
        preview["merge_request"] = new Dictionary<string, object?> { ["iid"] = mergeRequest.GetValueOrDefault("iid"), ["title"] = mergeRequest.GetValueOrDefault("title"), ["state"] = mergeRequest.GetValueOrDefault("state") };
        if (!MutationEnabled(context)) return preview;
        var posted = GitLab("POST", $"/projects/{Uri.EscapeDataString(project.Id)}/merge_requests/{Uri.EscapeDataString(iid)}/notes", token, body);
        if (posted.ContainsKey("error"))
        {
            preview["execution_status"] = "backend_error";
            preview["gitlab_error"] = posted;
            return preview;
        }
        preview["execution_status"] = "completed";
        preview["approval_required"] = false;
        preview["mutation_performed"] = true;
        preview["posted_comment"] = new Dictionary<string, object?> { ["id"] = posted.GetValueOrDefault("id") };
        return preview;
    }

    private static Dictionary<string, object?> PreparePipelineTrigger(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, string token)
    {
        var project = ProjectMetadata(capability, plan, parameters, token);
        if (project.Error is not null) return project.Error;
        var body = new Dictionary<string, object?> { ["ref"] = Text(parameters.GetValueOrDefault("ref")), ["variables"] = ObjectMap(parameters.GetValueOrDefault("variables")), ["purpose"] = Text(parameters.GetValueOrDefault("pipeline_purpose")) };
        return WritePreview(capability, plan, "pipeline.trigger", $"/projects/{project.Id}/pipeline", body, project.Payload!);
    }

    private static Dictionary<string, object?> PrepareReleaseNotes(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, string token)
    {
        var project = ProjectMetadata(capability, plan, parameters, token);
        if (project.Error is not null) return project.Error;
        var range = Text(parameters.GetValueOrDefault("range"));
        return Result(capability, plan, "completed", new()
        {
            ["mutation_performed"] = false,
            ["result"] = new Dictionary<string, object?>
            {
                ["title"] = $"Release notes for {FirstNonEmpty(Text(project.Payload!.GetValueOrDefault("path_with_namespace")), project.Id)} {range}",
                ["audience"] = FirstNonEmpty(Text(parameters.GetValueOrDefault("audience")), "internal"),
                ["project"] = ProjectSummary(project.Payload!),
                ["range"] = range,
                ["sections"] = new[]
                {
                    new Dictionary<string, object?> { ["title"] = "Highlights", ["items"] = new[] { "Review bounded GitLab context before publishing release notes." } },
                    new Dictionary<string, object?> { ["title"] = "Governance", ["items"] = new[] { "This capability drafts content only and does not create a GitLab release." } },
                },
            },
        });
    }

    public static Dictionary<string, object?> GitLab(string method, string path, string token, Dictionary<string, object?>? body)
    {
        using var request = new HttpRequestMessage(new HttpMethod(method), $"{ApiBase()}{path}");
        request.Headers.Accept.ParseAdd("application/json");
        request.Headers.UserAgent.ParseAdd("anip-gitlab-fronting-showcase");
        request.Headers.Add("PRIVATE-TOKEN", token);
        if (body is not null) request.Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");
        var response = Http.Send(request);
        var text = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
        var payload = ParseGitLabPayload(text);
        if (!response.IsSuccessStatusCode) return new Dictionary<string, object?> { ["error"] = "gitlab_http_error", ["status"] = (int)response.StatusCode, ["detail"] = payload };
        return payload;
    }

    private static ProjectContext ProjectMetadata(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, string token)
    {
        var project = ProjectId(parameters);
        if (!ProjectAllowed(project)) return new ProjectContext(project, null, Restricted(capability, plan, project));
        var payload = GitLab("GET", $"/projects/{Uri.EscapeDataString(project)}", token, null);
        if (payload.ContainsKey("error")) return new ProjectContext(project, null, Result(capability, plan, "backend_error", new() { ["gitlab_error"] = payload }));
        return new ProjectContext(project, payload, null);
    }

    private static Dictionary<string, object?> WritePreview(Dictionary<string, object?> capability, Dictionary<string, object?> plan, string action, string path, Dictionary<string, object?> body, Dictionary<string, object?> projectPayload) =>
        Result(capability, plan, "prepared", new()
        {
            ["approval_required"] = true,
            ["mutation_performed"] = false,
            ["gitlab_action"] = action,
            ["gitlab_metadata"] = ProjectSummary(projectPayload),
            ["gitlab_request"] = new Dictionary<string, object?> { ["method"] = "POST", ["path"] = path, ["body"] = body },
            ["note"] = "Prepared a GitLab request payload. No GitLab mutation was performed.",
        });

    private static Dictionary<string, object?> ProjectSummary(Dictionary<string, object?> payload) =>
        new() { ["id"] = payload.GetValueOrDefault("id"), ["path_with_namespace"] = payload.GetValueOrDefault("path_with_namespace"), ["default_branch"] = payload.GetValueOrDefault("default_branch"), ["visibility"] = payload.GetValueOrDefault("visibility"), ["web_url"] = payload.GetValueOrDefault("web_url") };

    private static Dictionary<string, object?> Result(Dictionary<string, object?> capability, Dictionary<string, object?> plan, string status, Dictionary<string, object?> extra)
    {
        var result = new Dictionary<string, object?> { ["execution_status"] = status, ["capability_id"] = capability.GetValueOrDefault("capability_id"), ["selected_backend"] = plan.GetValueOrDefault("selected_binding"), ["semantic_input"] = plan.GetValueOrDefault("semantic_input"), ["backend_input_contract"] = plan.GetValueOrDefault("backend_input_contract") };
        foreach (var item in extra) result[item.Key] = item.Value;
        return result;
    }

    private static Dictionary<string, object?> Restricted(Dictionary<string, object?> capability, Dictionary<string, object?> plan, string project) =>
        Result(capability, plan, "restricted", new() { ["project_id"] = project, ["reason"] = "GitLab project is outside the configured ANIP project policy." });

    private static string ProjectId(Dictionary<string, object?> parameters)
    {
        var explicitId = Text(parameters.GetValueOrDefault("project_id"));
        if (!string.IsNullOrWhiteSpace(explicitId)) return explicitId;
        var ns = Text(parameters.GetValueOrDefault("namespace")).Trim('/');
        var project = Text(parameters.GetValueOrDefault("project")).Trim('/');
        return !string.IsNullOrWhiteSpace(ns) && !string.IsNullOrWhiteSpace(project) ? $"{ns}/{project}" : "";
    }

    private static bool ProjectAllowed(string project)
    {
        var key = project.ToLowerInvariant();
        var blocked = CsvEnv("ANIP_GITLAB_BLOCKED_PROJECTS");
        var allowed = CsvEnv("ANIP_GITLAB_ALLOWED_PROJECTS");
        return !blocked.Contains(key) && (allowed.Count == 0 || allowed.Contains(key));
    }

    private static bool MutationEnabled(InvocationContext context) =>
        Environment.GetEnvironmentVariable("ANIP_GITLAB_ALLOW_MUTATION") == "true" && !string.IsNullOrWhiteSpace(context?.ApprovalGrant);

    private static string ApiBase()
    {
        var raw = Environment.GetEnvironmentVariable("GITLAB_API_BASE") ?? "https://gitlab.com/api/v4";
        return raw.TrimEnd('/');
    }

    private static Dictionary<string, object?> ParseGitLabPayload(string text)
    {
        if (string.IsNullOrWhiteSpace(text)) return new();
        var trimmed = text.TrimStart();
        if (trimmed.StartsWith("[", StringComparison.Ordinal)) return new() { ["items"] = JsonSerializer.Deserialize<JsonElement>(text) };
        return JsonSerializer.Deserialize<Dictionary<string, object?>>(text) ?? new();
    }

    private static List<JsonElement> JsonList(object? value)
    {
        var result = new List<JsonElement>();
        if (value is JsonElement element && element.ValueKind == JsonValueKind.Array)
        {
            foreach (var item in element.EnumerateArray()) result.Add(item);
        }
        return result;
    }

    private static Dictionary<string, object?> ObjectMap(object? value)
    {
        if (value is Dictionary<string, object?> dict) return dict;
        if (value is JsonElement element && element.ValueKind == JsonValueKind.Object)
        {
            return JsonSerializer.Deserialize<Dictionary<string, object?>>(element.GetRawText()) ?? new();
        }
        return new();
    }

    private static object? Get(JsonElement element, string key) => element.TryGetProperty(key, out var value) ? value.ValueKind == JsonValueKind.String ? value.GetString() : value.ToString() : null;
    private static string FirstNonEmpty(params string[] values) => values.FirstOrDefault(value => !string.IsNullOrWhiteSpace(value)) ?? "";
    private static int BoundedLimit(object? value, int defaultValue, int maximum) => int.TryParse(Text(value), out var parsed) ? Math.Max(1, Math.Min(parsed, maximum)) : defaultValue;
    private static string Text(object? value) => value is null ? "" : value.ToString()!.Trim();
    private static List<string> CsvEnv(string name) => (Environment.GetEnvironmentVariable(name) ?? "").Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries).Select(item => item.ToLowerInvariant()).Distinct().ToList();
    private static List<string> StringList(object? value) => value is IEnumerable<object?> items ? items.Select(Text).Where(item => item.Length > 0).Distinct().ToList() : value is string text ? text.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries).Distinct().ToList() : [];

    private sealed record ProjectContext(string Id, Dictionary<string, object?>? Payload, Dictionary<string, object?>? Error);
}

File: examples/showcase/gitlab_fronting/custom-code-bundles/gitlab_fronting_java/src/main/java/{{ANIP_JAVA_PACKAGE_PATH}}/BackendAdapter.java
package {{ANIP_JAVA_PACKAGE_NAME}};

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import dev.anip.service.InvocationContext;

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@FunctionalInterface
public interface BackendAdapter {

    Map<String, Object> execute(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> adapterInput, InvocationContext context);

    static BackendAdapter defaultAdapter() {
        return new GitLabBackendAdapter();
    }
}

final class GitLabBackendAdapter implements BackendAdapter {
    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final HttpClient HTTP = HttpClient.newHttpClient();

    @Override
    public Map<String, Object> execute(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, InvocationContext context) {
        List<String> unresolved = stringList(plan.get("unresolved_required_backend_inputs"));
        if (!unresolved.isEmpty()) {
            return result(capability, plan, "backend_input_incomplete", record("unresolved_required_backend_inputs", unresolved));
        }
        String token = System.getenv().getOrDefault("GITLAB_TOKEN", "").trim();
        if (token.isBlank()) return result(capability, plan, "backend_error", record("gitlab_error", record("error", "missing_gitlab_token")));
        return switch (text(capability.get("capability_id"))) {
            case "gitlab.project.search_context" -> searchProjectContext(capability, plan, params, token);
            case "gitlab.issue.prepare" -> prepareOrCreateIssue(capability, plan, params, token, context);
            case "gitlab.mr.comment.prepare" -> prepareMergeRequestComment(capability, plan, params, token, context);
            case "gitlab.pipeline.trigger.request" -> preparePipelineTrigger(capability, plan, params, token);
            case "gitlab.release_notes.prepare" -> prepareReleaseNotes(capability, plan, params, token);
            default -> result(capability, plan, "backend_execution_stub", record("note", "No GitLab custom handler is registered for this capability."));
        };
    }

    private static Map<String, Object> searchProjectContext(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String token) {
        String project = projectId(params);
        if (!projectAllowed(project)) return restricted(capability, plan, project);
        int limit = boundedLimit(params.get("limit"), 20, 50);
        String query = text(params.get("query"));
        Map<String, Object> issues = gitlab("GET", "/projects/" + urlPath(project) + "/issues?search=" + url(query) + "&per_page=" + limit, token, null);
        Map<String, Object> mrs = gitlab("GET", "/projects/" + urlPath(project) + "/merge_requests?search=" + url(query) + "&per_page=" + limit, token, null);
        if (issues.containsKey("error")) return result(capability, plan, "backend_error", record("gitlab_error", issues));
        if (mrs.containsKey("error")) return result(capability, plan, "backend_error", record("gitlab_error", mrs));
        List<Map<String, Object>> items = new ArrayList<>();
        for (Map<String, Object> item : mapList(issues.get("items"))) {
            if (items.size() >= limit) break;
            items.add(record("kind", "issue", "iid", item.get("iid"), "title", item.get("title"), "state", item.get("state"), "web_url", item.get("web_url")));
        }
        for (Map<String, Object> item : mapList(mrs.get("items"))) {
            if (items.size() >= limit) break;
            items.add(record("kind", "merge_request", "iid", item.get("iid"), "title", item.get("title"), "state", item.get("state"), "web_url", item.get("web_url")));
        }
        return result(capability, plan, "completed", record("gitlab_query", query, "result", record("items", items, "count", items.size(), "project_id", project)));
    }

    private static Map<String, Object> prepareOrCreateIssue(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String token, InvocationContext context) {
        ProjectContext project = projectMetadata(capability, plan, params, token);
        if (project.error != null) return project.error;
        Map<String, Object> body = record("title", text(params.get("title")), "description", firstNonBlank(params.get("body"), params.get("description")));
        List<String> labels = stringList(params.get("labels"));
        if (!labels.isEmpty()) body.put("labels", String.join(",", labels));
        Map<String, Object> preview = writePreview(capability, plan, "issues.create", "/projects/" + project.id + "/issues", body, project.payload);
        if (!mutationEnabled(context)) return preview;
        Map<String, Object> created = gitlab("POST", "/projects/" + urlPath(project.id) + "/issues", token, body);
        if (created.containsKey("error")) {
            preview.put("execution_status", "backend_error");
            preview.put("gitlab_error", created);
            return preview;
        }
        preview.put("execution_status", "completed");
        preview.put("approval_required", false);
        preview.put("mutation_performed", true);
        preview.put("created_issue", record("iid", created.get("iid"), "web_url", created.get("web_url"), "state", created.get("state")));
        preview.put("note", "Created GitLab issue after the ANIP runtime validated and reserved an approval grant.");
        return preview;
    }

    private static Map<String, Object> prepareMergeRequestComment(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String token, InvocationContext context) {
        ProjectContext project = projectMetadata(capability, plan, params, token);
        if (project.error != null) return project.error;
        String iid = text(params.get("merge_request_iid"));
        Map<String, Object> mr = gitlab("GET", "/projects/" + urlPath(project.id) + "/merge_requests/" + urlPath(iid), token, null);
        if (mr.containsKey("error")) return result(capability, plan, "backend_error", record("gitlab_error", mr));
        Map<String, Object> body = record("body", ("[" + firstNonBlank(params.get("comment_purpose"), "triage_update") + "] " + text(params.get("context"))).trim());
        Map<String, Object> preview = writePreview(capability, plan, "merge_requests.createNote", "/projects/" + project.id + "/merge_requests/" + iid + "/notes", body, project.payload);
        preview.put("merge_request", record("iid", mr.get("iid"), "title", mr.get("title"), "state", mr.get("state")));
        if (!mutationEnabled(context)) return preview;
        Map<String, Object> posted = gitlab("POST", "/projects/" + urlPath(project.id) + "/merge_requests/" + urlPath(iid) + "/notes", token, body);
        if (posted.containsKey("error")) {
            preview.put("execution_status", "backend_error");
            preview.put("gitlab_error", posted);
            return preview;
        }
        preview.put("execution_status", "completed");
        preview.put("approval_required", false);
        preview.put("mutation_performed", true);
        preview.put("posted_comment", record("id", posted.get("id")));
        return preview;
    }

    private static Map<String, Object> preparePipelineTrigger(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String token) {
        ProjectContext project = projectMetadata(capability, plan, params, token);
        if (project.error != null) return project.error;
        Map<String, Object> body = record("ref", text(params.get("ref")), "variables", objectMap(params.get("variables")), "purpose", text(params.get("pipeline_purpose")));
        return writePreview(capability, plan, "pipeline.trigger", "/projects/" + project.id + "/pipeline", body, project.payload);
    }

    private static Map<String, Object> prepareReleaseNotes(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String token) {
        ProjectContext project = projectMetadata(capability, plan, params, token);
        if (project.error != null) return project.error;
        String range = text(params.get("range"));
        return result(capability, plan, "completed", record(
            "mutation_performed", false,
            "result", record(
                "title", "Release notes for " + firstNonBlank(project.payload.get("path_with_namespace"), project.id) + " " + range,
                "audience", firstNonBlank(params.get("audience"), "internal"),
                "project", projectSummary(project.payload),
                "range", range,
                "sections", List.of(
                    record("title", "Highlights", "items", List.of("Review bounded GitLab context before publishing release notes.")),
                    record("title", "Governance", "items", List.of("This capability drafts content only and does not create a GitLab release."))
                )
            )
        ));
    }

    private static ProjectContext projectMetadata(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String token) {
        String project = projectId(params);
        if (!projectAllowed(project)) return new ProjectContext(project, null, restricted(capability, plan, project));
        Map<String, Object> payload = gitlab("GET", "/projects/" + urlPath(project), token, null);
        if (payload.containsKey("error")) return new ProjectContext(project, null, result(capability, plan, "backend_error", record("gitlab_error", payload)));
        return new ProjectContext(project, payload, null);
    }

    static Map<String, Object> gitlab(String method, String path, String token, Map<String, Object> body) {
        try {
            HttpRequest.Builder builder = HttpRequest.newBuilder(URI.create(apiBase() + path))
                .header("Accept", "application/json")
                .header("Content-Type", "application/json")
                .header("PRIVATE-TOKEN", token)
                .header("User-Agent", "anip-gitlab-fronting-showcase");
            if ("POST".equals(method)) builder.POST(HttpRequest.BodyPublishers.ofString(MAPPER.writeValueAsString(body == null ? Map.of() : body)));
            else builder.GET();
            HttpResponse<String> response = HTTP.send(builder.build(), HttpResponse.BodyHandlers.ofString());
            Map<String, Object> payload = response.body().isBlank() ? new LinkedHashMap<>() : parseGitLabBody(response.body());
            if (response.statusCode() < 200 || response.statusCode() >= 300) return record("error", "gitlab_http_error", "status", response.statusCode(), "detail", payload);
            return payload;
        } catch (Exception exc) {
            return record("error", "gitlab_http_error", "detail", exc.toString());
        }
    }

    private static Map<String, Object> parseGitLabBody(String body) throws Exception {
        String trimmed = body.trim();
        if (trimmed.startsWith("[")) return record("items", MAPPER.readValue(trimmed, new TypeReference<List<Map<String, Object>>>() {}));
        return MAPPER.readValue(trimmed, new TypeReference<>() {});
    }

    private static Map<String, Object> writePreview(Map<String, Object> capability, Map<String, Object> plan, String action, String path, Map<String, Object> body, Map<String, Object> projectPayload) {
        return result(capability, plan, "prepared", record(
            "approval_required", true,
            "mutation_performed", false,
            "gitlab_action", action,
            "gitlab_metadata", projectSummary(projectPayload),
            "gitlab_request", record("method", "POST", "path", path, "body", body),
            "note", "Prepared a GitLab request payload. No GitLab mutation was performed."
        ));
    }

    private static Map<String, Object> projectSummary(Map<String, Object> payload) {
        return record("id", payload.get("id"), "path_with_namespace", payload.get("path_with_namespace"), "default_branch", payload.get("default_branch"), "visibility", payload.get("visibility"), "web_url", payload.get("web_url"));
    }

    private static Map<String, Object> result(Map<String, Object> capability, Map<String, Object> plan, String status, Map<String, Object> extra) {
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("execution_status", status);
        result.put("capability_id", capability.get("capability_id"));
        result.put("selected_backend", plan.get("selected_binding"));
        result.put("semantic_input", plan.get("semantic_input"));
        result.put("backend_input_contract", plan.get("backend_input_contract"));
        result.putAll(extra);
        return result;
    }

    private static Map<String, Object> restricted(Map<String, Object> capability, Map<String, Object> plan, String project) {
        return result(capability, plan, "restricted", record("project_id", project, "reason", "GitLab project is outside the configured ANIP project policy."));
    }

    private static String projectId(Map<String, Object> params) {
        String explicit = text(params.get("project_id"));
        if (!explicit.isBlank()) return explicit;
        String namespace = text(params.get("namespace")).replaceAll("^/+|/+$", "");
        String project = text(params.get("project")).replaceAll("^/+|/+$", "");
        return !namespace.isBlank() && !project.isBlank() ? namespace + "/" + project : "";
    }

    private static boolean projectAllowed(String project) {
        String key = project.toLowerCase();
        List<String> blocked = csvEnv("ANIP_GITLAB_BLOCKED_PROJECTS");
        List<String> allowed = csvEnv("ANIP_GITLAB_ALLOWED_PROJECTS");
        return !blocked.contains(key) && (allowed.isEmpty() || allowed.contains(key));
    }

    private static boolean mutationEnabled(InvocationContext context) {
        String grant = context == null ? "" : text(context.getApprovalGrant());
        return "true".equals(System.getenv().getOrDefault("ANIP_GITLAB_ALLOW_MUTATION", "")) && !grant.isBlank();
    }

    private static String apiBase() {
        String raw = System.getenv().getOrDefault("GITLAB_API_BASE", "https://gitlab.com/api/v4").trim();
        return raw.endsWith("/") ? raw.substring(0, raw.length() - 1) : raw;
    }

    private static List<String> csvEnv(String name) {
        List<String> result = new ArrayList<>();
        for (String item : System.getenv().getOrDefault(name, "").split(",")) {
            String value = item.trim().toLowerCase();
            if (!value.isBlank() && !result.contains(value)) result.add(value);
        }
        return result;
    }

    private static int boundedLimit(Object value, int defaultValue, int max) {
        try {
            return Math.max(1, Math.min(Integer.parseInt(text(value)), max));
        } catch (Exception ignored) {
            return defaultValue;
        }
    }

    private static List<String> stringList(Object value) {
        List<String> result = new ArrayList<>();
        if (value instanceof List<?> items) {
            for (Object item : items) if (!text(item).isBlank()) result.add(text(item));
        } else if (value instanceof String raw) {
            for (String part : raw.split(",")) if (!part.trim().isBlank()) result.add(part.trim());
        }
        return result;
    }

    @SuppressWarnings("unchecked")
    private static List<Map<String, Object>> mapList(Object value) {
        List<Map<String, Object>> result = new ArrayList<>();
        if (value instanceof List<?> items) for (Object item : items) if (item instanceof Map<?, ?> map) result.add((Map<String, Object>) map);
        return result;
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> objectMap(Object value) {
        return value instanceof Map<?, ?> map ? (Map<String, Object>) map : new LinkedHashMap<>();
    }

    private static String firstNonBlank(Object... values) {
        for (Object value : values) if (!text(value).isBlank()) return text(value);
        return "";
    }

    private static String text(Object value) {
        return value == null ? "" : value.toString().trim();
    }

    private static String url(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    private static String urlPath(String value) {
        return url(value).replace("+", "%20");
    }

    private static Map<String, Object> record(Object... pairs) {
        Map<String, Object> result = new LinkedHashMap<>();
        for (int index = 0; index + 1 < pairs.length; index += 2) result.put(String.valueOf(pairs[index]), pairs[index + 1]);
        return result;
    }

    private record ProjectContext(String id, Map<String, Object> payload, Map<String, Object> error) {}
}
